File: ios/SkyCast/Features/Forecast/ForecastScreen.swift
import SwiftUI

/// The Forecast tab, five days for the primary location, tappable through to a day detail.
struct ForecastScreen: View {
    @Environment(AppContainer.self) private var container
    @State private var viewModel: ForecastViewModel?
    @State private var selectedDay: DayRoute?

    var body: some View {
        Group {
            if let viewModel {
                ForecastContent(
                    state: viewModel.state,
                    onRefresh: { await viewModel.refresh() },
                    onDismissBanner: viewModel.dismissBanner,
                    onOpenDay: { forecast, day in
                        selectedDay = DayRoute(locationID: forecast.locationID, date: day.date)
                    }
                )
            } else {
                LoadingView()
            }
        }
        .navigationTitle("Forecast")
        .navigationDestination(item: $selectedDay) { route in
            DayDetailScreen(locationID: route.locationID, date: route.date)
        }
        .task {
            if viewModel == nil {
                viewModel = ForecastViewModel(
                    weatherRepository: container.weatherRepository,
                    locationRepository: container.locationRepository,
                    settingsStore: container.settingsStore
                )
                viewModel?.start()
            } else {
                // The primary location may have changed on the Locations tab while this screen
                // was off-screen.
                await viewModel?.reload()
            }
        }
        .onDisappear { viewModel?.stop() }
    }
}

/// A pushed day-detail destination. A dedicated value type rather than a tuple, because
/// `navigationDestination(item:)` needs `Hashable`.
struct DayRoute: Hashable, Identifiable {
    let locationID: Int64
    let date: Date

    var id: String {
        "\(locationID)-\(date.timeIntervalSince1970)"
    }
}

/// The stateless half.
struct ForecastContent: View {
    let state: ForecastUiState
    let onRefresh: () async -> Void
    let onDismissBanner: () -> Void
    let onOpenDay: (Forecast, ForecastDay) -> Void

    var body: some View {
        if state.showsFullScreenLoader {
            LoadingView(message: "Fetching the forecast…")
        } else if state.showsEmptyState {
            EmptyStateView(
                title: "No locations yet",
                message: "Add a place on the Locations tab and its five-day forecast appears here.",
                systemImage: "mappin.and.ellipse"
            )
        } else if state.showsFullScreenError, let error = state.error {
            ErrorView(error: error, onRetry: { Task { await onRefresh() } })
        } else if let forecast = state.forecast {
            list(forecast)
        } else {
            // Reachable only if the API returns a forecast with no days at all.
            EmptyStateView(
                title: "No forecast available",
                message: "OpenWeather returned no daily readings for this place.",
                systemImage: "calendar.badge.exclamationmark"
            )
        }
    }

    private func list(_ forecast: Forecast) -> some View {
        List {
            if state.showsStaleBanner {
                StaleDataBanner(
                    message: state.staleBannerMessage,
                    onRetry: { Task { await onRefresh() } },
                    onDismiss: onDismissBanner
                )
                // The banner supplies its own glass padding; a row inset on top of it would
                // read as a second, misaligned pane.
                .listRowInsets(EdgeInsets())
                .listRowBackground(Color.clear)
            }

            Section(forecast.locationName) {
                ForEach(forecast.days) { day in
                    Button {
                        onOpenDay(forecast, day)
                    } label: {
                        ForecastDayRow(
                            day: day,
                            timeZone: forecast.timeZone,
                            unit: state.preferences.temperatureUnit
                        )
                    }
                    // A plain button style inside a List keeps the row looking like a row; the
                    // default styling would tint every label blue.
                    .buttonStyle(.plain)
                }
            }
        }
        .refreshable { await onRefresh() }
        // Subtle here: enough that the tabs feel like one app, not so much that a five-day list
        // competes with the forecast for attention.
        .weatherBackground(
            condition: forecast.days.first?.condition ?? .unknown,
            isDaytime: true,
            intensity: .subtle
        )
    }
}

/// One day of the forecast.
struct ForecastDayRow: View {
    let day: ForecastDay
    /// The forecast location's zone. `day.date` is the start of that day *there*, so labelling it
    /// in the device's zone can name the wrong weekday entirely.
    let timeZone: TimeZone
    let unit: TemperatureUnit

    private var high: Int {
        Int(unit.convertFromCelsius(day.maxTemperatureCelsius).rounded())
    }

    private var low: Int {
        Int(unit.convertFromCelsius(day.minTemperatureCelsius).rounded())
    }

    private var rainPercent: Int {
        Int((day.precipitationProbability * 100).rounded())
    }

    private var dayLabel: String {
        // The zone goes in the initialiser. `.timeZone(_:)` on the style would instead add the
        // zone's *name* to the output, which is a different thing entirely.
        let style = Date.FormatStyle(timeZone: timeZone)
            .weekday(.wide)
            .day()
            .month(.abbreviated)
        return day.date.formatted(style)
    }

    var body: some View {
        HStack(spacing: Spacing.md) {
            ConditionBadge(
                condition: day.condition,
                // A forecast row summarises a whole day, so daytime artwork is the honest choice.
                isDaytime: true,
                size: forecastBadgeSize
            )

            VStack(alignment: .leading, spacing: Spacing.xxs) {
                Text(dayLabel)
                    .font(.headline)
                Text(day.description)
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }

            Spacer(minLength: Spacing.sm)

            VStack(alignment: .trailing, spacing: Spacing.xxs) {
                Text("\(high)° / \(low)°")
                    .font(.headline)
                if rainPercent > 0 {
                    Label("\(rainPercent)%", systemImage: "drop.fill")
                        .font(.caption)
                        .foregroundStyle(Color.skyAccent)
                        .labelStyle(.titleAndIcon)
                }
            }
        }
        .padding(.vertical, Spacing.xs)
        // One announcement per row; otherwise VoiceOver reads the date, two bare numbers and a
        // percentage as four disconnected fragments.
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(announcement)
        .accessibilityHint("Shows the hour-by-hour breakdown")
    }

    private var announcement: String {
        var text = "\(dayLabel), \(day.description), high \(high)\(unit.symbol), "
            + "low \(low)\(unit.symbol)"
        if rainPercent > 0 {
            text += ", \(rainPercent) percent chance of rain"
        }
        return text
    }

    /// Smaller than the Home hero badge, 44 pt is also the minimum touch target, so the row
    /// stays tappable at its natural height.
    private let forecastBadgeSize = TouchTarget.minimum
}

// MARK: - Previews

#Preview("Loading") {
    NavigationStack {
        ForecastContent(
            state: ForecastUiState(isLoading: true),
            onRefresh: {},
            onDismissBanner: {},
            onOpenDay: { _, _ in }
        )
        .navigationTitle("Forecast")
    }
}

#Preview("Empty") {
    NavigationStack {
        ForecastContent(
            state: ForecastUiState(hasNoLocation: true),
            onRefresh: {},
            onDismissBanner: {},
            onOpenDay: { _, _ in }
        )
        .navigationTitle("Forecast")
    }
}

#Preview("Offline, no cache") {
    NavigationStack {
        ForecastContent(
            state: ForecastUiState(error: .offline),
            onRefresh: {},
            onDismissBanner: {},
            onOpenDay: { _, _ in }
        )
        .navigationTitle("Forecast")
    }
}

File: android/app/src/main/java/com/nauhaan/skycast/ui/forecast/ForecastScreen.kt
package com.nauhaan.skycast.ui.forecast

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AddLocationAlt
import androidx.compose.material.icons.filled.CloudOff
import androidx.compose.material.icons.filled.WaterDrop
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.nauhaan.skycast.R
import com.nauhaan.skycast.core.designsystem.component.EmptyStateView
import com.nauhaan.skycast.core.designsystem.component.ErrorView
import com.nauhaan.skycast.core.designsystem.component.LoadingView
import com.nauhaan.skycast.core.designsystem.component.StaleDataBanner
import com.nauhaan.skycast.core.designsystem.component.WeatherConditionBadge
import com.nauhaan.skycast.core.designsystem.theme.Spacing
import com.nauhaan.skycast.domain.model.ForecastDay
import com.nauhaan.skycast.domain.model.TemperatureUnit
import com.nauhaan.skycast.ui.common.toPresentation
import java.time.format.DateTimeFormatter
import kotlin.math.roundToInt

/** The Forecast tab, five days for the primary location, tappable through to a day detail. */
@Composable
fun ForecastScreen(
    onNavigateToDayDetail: (locationId: Long, epochDay: Long) -> Unit,
    modifier: Modifier = Modifier,
    viewModel: ForecastViewModel = hiltViewModel(),
) {
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()

    ForecastContent(
        uiState = uiState,
        onRefresh = viewModel::refresh,
        onOpenDay = onNavigateToDayDetail,
        modifier = modifier,
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun ForecastContent(
    uiState: ForecastUiState,
    onRefresh: () -> Unit,
    onOpenDay: (locationId: Long, epochDay: Long) -> Unit,
    modifier: Modifier = Modifier,
) {
    when {
        uiState.showsFullScreenLoader -> LoadingView(modifier = modifier)

        uiState.showsEmptyState -> EmptyStateView(
            title = stringResource(R.string.today_empty_title),
            message = stringResource(R.string.today_empty_message),
            icon = Icons.Filled.AddLocationAlt,
            modifier = modifier,
        )

        uiState.showsFullScreenError -> {
            val presentation = requireNotNull(uiState.error).toPresentation()
            ErrorView(
                title = stringResource(presentation.titleRes),
                message = stringResource(presentation.messageRes),
                icon = Icons.Filled.CloudOff,
                onRetry = onRefresh.takeIf { presentation.isRetryable },
                modifier = modifier,
            )
        }

        else -> PullToRefreshBox(
            isRefreshing = uiState.isRefreshing,
            onRefresh = onRefresh,
            modifier = modifier.fillMaxSize(),
        ) {
            LazyColumn {
                if (uiState.showsStaleBanner) {
                    item {
                        val message = uiState.error
                            ?.let { stringResource(it.toPresentation().messageRes) }
                            ?: stringResource(R.string.banner_data_may_be_out_of_date)
                        StaleDataBanner(message = message, onRetry = onRefresh)
                    }
                }
                uiState.forecast?.let { forecast ->
                    item {
                        Text(
                            text = forecast.locationName,
                            style = MaterialTheme.typography.titleLargeEmphasized,
                            modifier = Modifier.padding(Spacing.md),
                        )
                    }
                    items(forecast.days, key = { it.date.toEpochDay() }) { day ->
                        ForecastDayRow(
                            day = day,
                            unit = uiState.preferences.temperatureUnit,
                            onClick = { onOpenDay(forecast.locationId, day.date.toEpochDay()) },
                        )
                        HorizontalDivider()
                    }
                }
            }
        }
    }
}

/**
 * One day of the forecast.
 *
 * `clearAndSetSemantics` merges the row into a single announcement, otherwise TalkBack reads
 * the date, two bare numbers and a percentage as four disconnected fragments.
 */
@Composable
private fun ForecastDayRow(
    day: ForecastDay,
    unit: TemperatureUnit,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val high = unit.convertFromCelsius(day.maxTemperatureCelsius).roundToInt()
    val low = unit.convertFromCelsius(day.minTemperatureCelsius).roundToInt()
    val rain = (day.precipitationProbability * PERCENT).roundToInt()
    // Weekday first, and no year: the question a forecast row answers is "what about Thursday?",
    // so the weekday is the useful part and "2026" is noise. Matches the iOS row.
    val dayLabel = day.date.format(DayLabelFormat)
    val announcement = stringResource(
        R.string.forecast_day_accessibility,
        dayLabel,
        day.description,
        high,
        unit.symbol,
        low,
        rain,
    )

    Row(
        modifier = modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .clearAndSetSemantics { contentDescription = announcement }
            .padding(horizontal = Spacing.md, vertical = Spacing.sm),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        WeatherConditionBadge(
            condition = day.condition,
            // Forecast rows summarise a whole day, so daytime artwork is the honest choice.
            isDaytime = true,
            size = BadgeSize,
        )
        Column(
            modifier = Modifier
                .weight(1f)
                .padding(start = Spacing.md),
        ) {
            Text(text = dayLabel, style = MaterialTheme.typography.titleMedium)
            Text(
                text = day.description,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        if (rain > 0) {
            Icon(
                imageVector = Icons.Filled.WaterDrop,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.primary,
                modifier = Modifier.size(16.dp),
            )
            Text(
                text = "$rain%",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.primary,
                modifier = Modifier.padding(end = Spacing.sm),
            )
        }
        Text(
            text = stringResource(R.string.forecast_high_low, high, low),
            style = MaterialTheme.typography.titleMedium,
            textAlign = TextAlign.End,
        )
    }
}

private val DayLabelFormat: DateTimeFormatter = DateTimeFormatter.ofPattern("EEEE d MMM")

private const val PERCENT = 100
private val BadgeSize = 44.dp
